modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,410 @@
+                        processDeployRequest((DeployUnitRequest) message, senderConsistentId, correlationId);
+                    } else if (message instanceof UndeployUnitRequest) {
+                        processUndeployRequest((UndeployUnitRequest) message, senderConsistentId, correlationId);
+                    }
+                });
+    }
+
+    private void processDeployRequest(DeployUnitRequest executeRequest, String senderConsistentId, long correlationId) {
+        doDeploy(executeRequest).whenComplete((success, throwable) -> {
+            DeployUnitResponseBuilder builder = DeployUnitResponseImpl.builder();
+            if (throwable != null) {
+                builder.error(throwable);
+            }
+            clusterService.messagingService().respond(senderConsistentId,
+                    builder.build(), correlationId);
+        });
+    }
+
+    private void processUndeployRequest(UndeployUnitRequest executeRequest, String senderConsistentId, long correlationId) {
+        try {
+            Path unitPath = unitsFolder
+                    .resolve(executeRequest.id())
+                    .resolve(executeRequest.version());
+
+            Files.walkFileTree(unitPath, new SimpleFileVisitor<>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    Files.delete(file);
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    Files.delete(dir);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        } catch (IOException e) {
+            LOG.error("Failed to undeploy unit " + executeRequest.id() + ":" + executeRequest.version(), e);
+            clusterService.messagingService()
+                    .respond(senderConsistentId, UndeployUnitResponseImpl.builder().error(e).build(), correlationId);
+            return;
+        }
+
+        clusterService.messagingService()
+                .respond(senderConsistentId, UndeployUnitResponseImpl.builder().build(), correlationId);
+    }
+
+    private CompletableFuture<Boolean> doDeploy(DeployUnitRequest executeRequest) {
+        String id = executeRequest.id();
+        String version = executeRequest.version();
+        try {
+            Path unitPath = unitsFolder
+                    .resolve(executeRequest.id())
+                    .resolve(executeRequest.version())
+                    .resolve(executeRequest.unitName());
+            Path unitPathTmp = unitPath.resolveSibling(unitPath.getFileName() + ".tmp");
+            Files.createDirectories(unitPathTmp.getParent());
+            Files.write(unitPathTmp, executeRequest.unitContent(),
+                    StandardOpenOption.CREATE, StandardOpenOption.SYNC, StandardOpenOption.TRUNCATE_EXISTING);
```

Comment:
DSYNC should be enough

Reply:
Ok. Probably DSYNC won't even work with TRUNCATE, as TRUNCATE changes file's attribute